File: atcoder/ABC075/ABC075C.py
def dfs(v):
    visited[v]=True
    for v2 in range(n):
        if graph[v][v2]==False:
            continue
        if visited[v2]==True:
            continue
        dfs(v2)

n, m = map(int, input().split())

alis=[]
blis=[]
graph=[[False] * n for i in range(n)]

#隣接行列のセット
for i in range(m):
    a, b = map(int, input().split())
    graph[a-1][b-1]=True
    graph[b-1][a-1]=True
    alis.append(a-1)
    blis.append(b-1)

ans=0

for i in range(m):
    #1本辺を落とす
    graph[alis[i]][blis[i]]=False
    graph[blis[i]][alis[i]]=False
    visited=[False]*n
    dfs(0)
    bridge=False
    for j in range(n):
        if visited[j]==False:
            bridge=True
    if bridge:
        ans+=1
    graph[alis[i]][blis[i]]=True
    graph[blis[i]][alis[i]]=True

print(ans)

File: atcoder/ABC075/ABC075C.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <functional>
#include <stack>
#include <queue>
#include <string>
#include <map>
#include <unordered_map>
#include <bitset>
#include <numeric>


using namespace std;
typedef long long ll;
typedef pair<ll,ll> P;

const ll mod=1000000007ll;
const ll inf=1000000000000000ll;

const ll nmax=50;

ll d[nmax][nmax],a[nmax][nmax];
P edge[nmax];
ll n;
bool visited[nmax],allvisited;

void dfs(ll c){
    visited[c]=true;
    for(ll i=0;i<n;i++){
        if(d[c][i]==1&&!(visited[i])){
            dfs(i);
            }
        }
    }


int main(){
    ll m,a,b,ans;
    cin>>n>>m;
    ans=0;
    for(ll i=0;i<m;i++){
        cin>>a>>b;
        edge[i]=P(a-1,b-1);
    }
    for(ll ct=0;ct<n;ct++){
        for(ll ct2=0;ct2<n;ct2++){
            if(ct==ct2)d[ct][ct2]=0;
            else d[ct][ct2]=inf;
        }
    }
    for(ll ct=0;ct<m;ct++){
        d[edge[ct].first][edge[ct].second]=1;
        d[edge[ct].second][edge[ct].first]=1;
    }

    for(ll i=0;i<m;i++){
        d[edge[i].first][edge[i].second]=inf;
        d[edge[i].second][edge[i].first]=inf;

        for(ll j=0;j<n;j++)visited[j]=false;
        dfs(0);

        allvisited=true;
        for(ll i=0;i<n;i++){
            if(!(visited[i])){
                allvisited=false;
                break;
            }
        }
        if(!(allvisited)) ans++;


        d[edge[i].first][edge[i].second]=1;
        d[edge[i].second][edge[i].first]=1;
    }
    cout<<ans<<endl;
    return 0;
}
